sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -652,10 +652,19 @@ object DataSourceStrategy {
  */
 object PushableColumn {
   def unapply(e: Expression): Option[String] = {
-    def helper(e: Expression) = e match {
-      case a: Attribute => Some(a.name)
+    val nestedPredicatePushdownEnabled = SQLConf.get.nestedPredicatePushdownEnabled
+    import org.apache.spark.sql.connector.catalog.CatalogV2Implicits.MultipartIdentifierHelper
+    def helper(e: Expression): Option[Seq[String]] = e match {
+      case a: Attribute =>
+        if (nestedPredicatePushdownEnabled || !a.name.contains(".")) {
+          Some(Seq(a.name))
+        } else {
+          None
+        }
+      case s: GetStructField if nestedPredicatePushdownEnabled =>
+        helper(s.child).map(_ :+ s.childSchema(s.ordinal).name)
       case _ => None
     }
-    helper(e)
+    helper(e).map(_.quoted)
```

Comment:
Sorry, for my late review. This sounds an API breaking change.

Reply:
Can we limit this change to some specific data sources? For example, parquet only?

Reply:
I am afraid it might break the released external third-party connectors that might not be able to handle the quoted column names.

Reply:
Yes, this was pointed out and discussed at https://github.com/apache/spark/pull/27728#discussion_r397742247 and https://github.com/apache/spark/pull/27728#discussion_r397745115.
@HeartSaVioR is working on it against this JIRA SPARK-31365. I turned the JIRA to be a blocker for Spark 3.0.

Reply:
I'm sorry I've been working on my own task (prioritized) and am afraid I can't pick up this soon. Please take this over if anyone has the idea of implementing this.

Reply:
@viirya Are you interested in this follow up?

Reply:
I can try looking at this this week. If anyone picks it up before me, I'm also ok.

Reply:
Thank you for fixing this 3.0 blocker